python/ray/experimental/data/read_api.py

```diff
@@ -17,12 +17,30 @@
 def autoinit_ray(f):
     def wrapped(*a, **kw):
         if not ray.is_initialized():
-            ray.client.connect()
+            ray.client().connect()
         return f(*a, **kw)
 
     return wrapped
 
 
+def _parse_paths(paths: Union[str, List[str]]
+                 ) -> Tuple["pyarrow.fs.FileSystem", Union[str, List[str]]]:
+    from pyarrow import fs
+    if isinstance(paths, str):
+        return fs.FileSystem.from_uri(paths)
+    else:
+        if len(paths) == 0:
+            raise ValueError("No data provided")
+        parsed_results = [fs.FileSystem.from_uri(path) for path in paths]
+        filesystem = parsed_results[0][0]
+        paths = []
+        for (tmp_fs, path) in parsed_results:
+            if type(tmp_fs) != type(filesystem):
+                raise ValueError("Paths contains multiple filesystem")
+            paths.append(path)
```

Comment:
Supernit: This check might be simpler and provide a better error message with something like the following:
```python
fses, paths = zip(*parsed_results)
unique_fses = set(map(type, fses))
if len(fses) > 1:
    raise ValueError(
        f"When specifying multiple paths, each path must have the "
        f"same filesystem, but found: {fses}")
return fses[0], paths
```
I find this simpler, but others may like the explicit for loop rather than Python's functional tooling.